src/api/v1/users/register/volunteers.ts

```diff
@@ -51,23 +53,77 @@ export default [
       response: { schema: response },
     },
     handler: async (request: VolunteerRegisterRequest, h: Hapi.ResponseToolkit) => {
-      const { server: { app: { knex } }, payload } = request;
+      const {
+        server: { app: { knex, EmailService } },
+        payload,
+      } = request;
       const { email, role, organisationId, adminCode } = payload;
       /*
        * Preliminaries
        */
-      // check user doesn't already exist
-
-      if (await Users.exists(knex, { where: { email } })) {
-        return Boom.conflict('User with this e-mail already registered');
-      }
 
       const communityBusiness = await CommunityBusinesses
-      .getOne(knex, { where: { id: organisationId, deletedAt: null } });
+        .getOne(knex, { where: { id: organisationId, deletedAt: null } });
 
       // check organisation exists
       if (!communityBusiness) return Boom.badRequest('Unrecognised organisation');
 
+
+      // Check user exists
+      if (await Users.exists(knex, { where: { email } })) {
+        // VOLUNTEER_ADMIN cannot have a second role
+        if (role === RoleEnum.VOLUNTEER_ADMIN) {
+          return Boom.conflict('User with this e-mail already registered');
+        }
+        // Registering second role
+        const user = await Users.getOne(knex, { where: { email } });
+        const currentRoles = await Roles.fromUser(knex, user);
+        // check volunteer account doesn already exist
+        if (currentRoles.some((x) =>
+          x.role === RoleEnum.VOLUNTEER || x.role === RoleEnum.VOLUNTEER_ADMIN)) {
+          throw Boom.conflict('Volunteer with this e-mail already registered');
+        }
+
+        // Currently not supporting roles at different cbs
+        const orgMatchesCurrentRoles = currentRoles.every((x) =>
+              x.organisationId === organisationId);
+
+        if (!orgMatchesCurrentRoles) {
+          throw Boom.conflict(
+            'User with this e-mail already registered at another Community Business');
+        }
+
+        const { token } = await Users.createConfirmAddRoleToken(knex, user);
+
+        try {
+          await EmailService.send({
```

Comment:
I think there's room for an intermediate abstraction that sits between the generic parts of the `EmailService` and the handler (application-specific logic), in order to make the handler more declarative.

Something like:
```
EmailService.confirmAddRole(user, cb, token, role)
```

Reply:
To be completed in separate refactor PR